flare/framework/env.py

```diff
@@ -1,25 +1,163 @@
 from scipy import misc
+from abc import ABCMeta, abstractmethod
+from py_simulator import Simulator
 """
 A file for wrapping different environments to have a unified API interface
 used by Agent's control flow.
 """
 
 
-class GymEnv(object):
+class Env(object):
+    """
+    An abstract class for environment. A new environment should inherit this
+    class
+    """
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def reset(self):
+        """
+        reset the environment and return the initial observation
+        """
+        pass
+
+    @abstractmethod
+    def step(self, actions, actrep):
+        """
+        Given a list of ordered actions, forward the environment actrep step.
+        The output should be a list of next observations, a list of rewards,
+        and the next game over.
+        """
+        pass
+
+    @abstractmethod
+    def observation_dims(self):
+        """
+        Return a list of lists as observation dimensions, each list for one 
+        observation.
+        Each list contains the dimension numbers of that input.
+        """
+        pass
+
+    @abstractmethod
+    def action_dims(self):
+        """
+        Return a list of integers as action dimensions, each integer for an action.
+        For each integer,
+        if the corresponding action is discrete, then it means
+        the total number of actions;
+        if continous, then it means the length of the action vector.
+        if language, then it means the cardinality of the dictionary
+        """
+        pass
+
+    @abstractmethod
+    def time_out(self):
+        """
+        Return a boolean of whether the env has timed out 
+        """
+        pass
+
+
+class XWorldEnv(Env):
+    """
+    A wrapper for XWorld env.
+    """
+
+    def __init__(self, game_name, contexts=1, options=None, dict=None):
+        """
+        options: see test_xworld3d.py in XWorld for an example of options
+        dict: dictionary for the langauge part. key=id, value=word
+        """
+        self.env = Simulator.create(game_name, options)
+        assert contexts >= 1, "contexts must be a positive number!"
+        self.contexts = contexts
+        self.buf = None
+        self.dict_id2w = dict
+        self.dict_w2id = {v: k for k, v in dict.iteritems()}
+        self.height, self.width, self.channels = \
+                                        self.env.get_screen_out_dimensions()
+
+    def preprocess_observation(self, ob):
+        screen = ob['screen'].reshape([self.channels, self.height, self.width])
+        sentence = [
+            np.array(self.dict_w2id(word))
+            for word in ob['sentence'].split(" ")
+        ]
+
+        return screen, sentence
+
+    def reset(self):
+        ## should return a list of observations
+        self.env.reset_game()
+        init_ob = self.env.get_state()
+        screen, sentence = self.preprocess_observation(init_ob)
+
+        self.buf = [np.zeros(screen.shape).astype(screen.dtype) \
```

Comment:
xworld by default has an option of 'context'. So probably you don't need to maintain here.